docs(measurements): add code blocks to introductory tutorial (#1120)
vprusso/toqito: docs/intro_tutorial.rst

    >>> non_meas_1 = np.array([[1, 2], [3, 4]])
    >>> non_meas_2 = np.array([[5, 6], [7, 8]])
    >>> non_meas = [non_meas_1, non_meas_2]
    >>> is_povm(non_meas)
    False

Consider the following state:

.. math::
    u = \frac{1}{\sqrt{3}} e_0 + \sqrt{\frac{2}{3}} e_1

where we define :math:`u u^* = \rho \in \text{D}(\mathcal{X})`.

Define measurement operators

.. math::
    P_0 = e_0 e_0^* \quad \text{and} \quad P_1 = e_1 e_1^*.

.. code-block:: python

    >>> from toqito.states import basis
    >>> from toqito.measurement_ops import measure
    >>> import numpy as np
    >>> e_0, e_1 = basis(2, 0), basis(2, 1)
    >>>
    >>> u = 1/np.sqrt(3) * e_0 + np.sqrt(2/3) * e_1
    >>> rho = u @ u.conj().T
    >>>
    >>> proj_0 = e_0 @ e_0.conj().T
    >>> proj_1 = e_1 @ e_1.conj().T

Then the probability of obtaining outcome :math:`0` is given by

.. math::
    \langle P_0, \rho \rangle = \frac{1}{3}.

.. code-block:: python

    >>> measure(proj_0, rho)
    0.3333333333333334

Similarly, the probability of obtaining outcome :math:`1` is given by

.. math::
    \langle P_1, \rho \rangle = \frac{2}{3}.

.. code-block:: python

    >>> measure(proj_1, rho)
    0.6666666666666666

> Consider the collection of trine states.

Reviewer: We should likely add some type of description and reference about the "pretty good measurement". Otherwise this example seems like it's just kind of hanging out there in limbo.

.. math::
    u_0 = |0\rangle, \quad
    u_1 = -\frac{1}{2}\left(|0\rangle + \sqrt{3}|1\rangle\right), \quad \text{and} \quad
    u_2 = -\frac{1}{2}\left(|0\rangle - \sqrt{3}|1\rangle\right).

.. code-block:: python

    >>> from toqito.states import trine
    >>> from toqito.measurements import pretty_good_measurement
    >>>
    >>> states = trine()
    >>> probs = [1 / 3, 1 / 3, 1 / 3]
    >>> pgm = pretty_good_measurement(states, probs)
    >>> pgm
    [array([[0.66666667, 0.        ],
           [0.        , 0.        ]]), array([[0.16666667, 0.28867513],
           [0.28867513, 0.5       ]]), array([[ 0.16666667, -0.28867513],
           [-0.28867513,  0.5       ]])]

> Consider the collection of trine states.

Reviewer: We might want to have a different lead-in sentence here (as we already say "Consider the collection of trine states." up above. Maybe something like:

```
Similarly, we can consider so-called "pretty bad measurement" on the set of trine states
```

We should also provide a reference here to the "pretty bad measurement" paper (which is in the `refs.bib` file already). Feel free to modify/tweak/add to the sentence above as well.

.. math::
    u_0 = |0\rangle, \quad
    u_1 = -\frac{1}{2}\left(|0\rangle + \sqrt{3}|1\rangle\right), \quad \text{and} \quad
    u_2 = -\frac{1}{2}\left(|0\rangle - \sqrt{3}|1\rangle\right).

.. code-block:: python

    >>> from toqito.states import trine
    >>> from toqito.measurements import pretty_bad_measurement
    >>>
    >>> states = trine()
    >>> probs = [1 / 3, 1 / 3, 1 / 3]

>     >>> pgm = pretty_bad_measurement(states, probs)

Reviewer: We likely shouldn't name this variable `pgm` (probably `pbm` instead).

where :math:`\Sigma` represents a set of measurement outcomes and where
:math:`\mu(a)` represents the measurement operator associated with outcome
:math:`a \in \Sigma`.

> 

Reviewer 2: Add subsection: POVM

Consider the following matrices:

.. math::
    M_0 =
    \begin{pmatrix}
        1 & 0 \\
        0 & 0
    \end{pmatrix}
    \quad \text{and} \quad
    M_1 =
    \begin{pmatrix}
        0 & 0 \\
        0 & 1
    \end{pmatrix}.

Our function indicates that this set of operators constitute a set of

> POVMs.

Reviewer 2: briefly define what a POVM is and why our function expects it to be a POVM.

    >>> from toqito.measurement_props import is_povm
    >>> import numpy as np
    >>> meas_1 = np.array([[1, 0], [0, 0]])
    >>> meas_2 = np.array([[0, 0], [0, 1]])
    >>> meas = [meas_1, meas_2]
    >>> is_povm(meas)
    True

> 

Reviewer 2: Add subsection: Random POVM

We may also use the :code:`random_povm` function from :code:`toqito`, and can verify that a

> randomly generated set satisfies the criteria for being a POVM set.

Reviewer 2: briefly define what a POVM set is...

    >>> from toqito.measurement_props import is_povm
    >>> from toqito.rand import random_povm
    >>> import numpy as np
    >>> dim, num_inputs, num_outputs = 2, 2, 2
    >>> measurements = random_povm(dim, num_inputs, num_outputs)
    >>> is_povm([measurements[:, :, 0, 0], measurements[:, :, 0, 1]])
    True

Alternatively, the following matrices

.. math::
    M_0 =
    \begin{pmatrix}
        1 & 2 \\
        3 & 4
    \end{pmatrix}
    \quad \text{and} \quad
    M_1 =
    \begin{pmatrix}
        5 & 6 \\
        7 & 8
    \end{pmatrix},

do not constitute a POVM set.

    >>> from toqito.measurement_props import is_povm
    >>> import numpy as np
    >>> non_meas_1 = np.array([[1, 2], [3, 4]])
    >>> non_meas_2 = np.array([[5, 6], [7, 8]])
    >>> non_meas = [non_meas_1, non_meas_2]
    >>> is_povm(non_meas)
    False

> 

Reviewer 2: Add a subsection

Consider the following state:

.. math::
    u = \frac{1}{\sqrt{3}} e_0 + \sqrt{\frac{2}{3}} e_1

where we define :math:`u u^* = \rho \in \text{D}(\mathcal{X})`.

Define measurement operators

.. math::

>     P_0 = e_0 e_0^* \quad \text{and} \quad P_1 = e_1 e_1^*.

Reviewer 2: provide the latex equation for `e_0` and `e_1`.

.. code-block:: python

    >>> from toqito.states import basis
    >>> from toqito.measurement_ops import measure
    >>> import numpy as np
    >>> e_0, e_1 = basis(2, 0), basis(2, 1)
    >>>
    >>> u = 1/np.sqrt(3) * e_0 + np.sqrt(2/3) * e_1
    >>> rho = u @ u.conj().T
    >>>
    >>> proj_0 = e_0 @ e_0.conj().T
    >>> proj_1 = e_1 @ e_1.conj().T

Then the probability of obtaining outcome :math:`0` is given by

.. math::
    \langle P_0, \rho \rangle = \frac{1}{3}.

.. code-block:: python

    >>> measure(proj_0, rho)
    0.3333333333333334

Similarly, the probability of obtaining outcome :math:`1` is given by

.. math::
    \langle P_1, \rho \rangle = \frac{2}{3}.

.. code-block:: python

    >>> measure(proj_1, rho)
    0.6666666666666666

> 

Reviewer 2: add a subsection

Consider "pretty good measurement" on the set of trine states.

> The pretty good measurement also known as the "square root measurement" is a set of POVMs :math:`(G_1, \ldots, G_n)` defined as

Reviewer 2: ```suggestion
The pretty good measurement (PGM), also known as the "square root measurement" is a set of POVMs :math:`(G_1, \ldots, G_n)` defined as
```

.. math::
    G_i = P^{-1/2} \left(p_i \rho_i\right) P^{-1/2} \quad \text{where} \quad P = \sum_{i=1}^n p_i \rho_i.

This measurement was initially defined in :cite:`Hughston_1993_Complete` and has found applications in quantum state discrimination tasks. 
While not always optimal, the PGM provides a reasonable measurement strategy that can be computed efficiently.

For example, consider the following trine states:

.. math::
    u_0 = |0\rangle, \quad
    u_1 = -\frac{1}{2}\left(|0\rangle + \sqrt{3}|1\rangle\right), \quad \text{and} \quad
    u_2 = -\frac{1}{2}\left(|0\rangle - \sqrt{3}|1\rangle\right).

.. code-block:: python

    >>> from toqito.states import trine
    >>> from toqito.measurements import pretty_good_measurement
    >>>
    >>> states = trine()
    >>> probs = [1 / 3, 1 / 3, 1 / 3]
    >>> pgm = pretty_good_measurement(states, probs)
    >>> pgm
    [array([[0.66666667, 0.        ],
           [0.        , 0.        ]]), array([[0.16666667, 0.28867513],
           [0.28867513, 0.5       ]]), array([[ 0.16666667, -0.28867513],
           [-0.28867513,  0.5       ]])]

> 

Reviewer 2: add a subsection

> Similarly, we can consider so called "pretty bad measurement" on the set of trine states :cite:`McIrvin_2024_Pretty`.

Reviewer 2: ```suggestion
Similarly, we can consider so-called "pretty bad measurement" (PBM) on the set of trine states :cite:`McIrvin_2024_Pretty`.
```

Be consistent with the formatting. Here you are using quotes, but for PGM you did not.

.. math::
    u_0 = |0\rangle, \quad
    u_1 = -\frac{1}{2}\left(|0\rangle + \sqrt{3}|1\rangle\right), \quad \text{and} \quad
    u_2 = -\frac{1}{2}\left(|0\rangle - \sqrt{3}|1\rangle\right).

.. code-block:: python

    >>> from toqito.states import trine
    >>> from toqito.measurements import pretty_bad_measurement
    >>>
    >>> states = trine()
    >>> probs = [1 / 3, 1 / 3, 1 / 3]
    >>> pbm = pretty_bad_measurement(states, probs)
    >>> pbm
    [array([[0.16666667, 0.        ],
           [0.        , 0.5       ]]), array([[ 0.41666667, -0.14433757],
           [-0.14433757,  0.25      ]]), array([[0.41666667, 0.14433757],

>            [0.14433757, 0.25      ]])]

Reviewer 2: If you are citing references, add a section to list all the references. take a look at the other tutorials/function docstrings.

where :math:`\Sigma` represents a set of measurement outcomes and where
:math:`\mu(a)` represents the measurement operator associated with outcome
:math:`a \in \Sigma`.

POVM
^^^^

> 

Reviewer 2: ```suggestion
POVM (Positive Operator-Valued Measure) is a set of positive operators that sum up to the identity. 
```

Consider the following matrices:

.. math::
    M_0 =
    \begin{pmatrix}
        1 & 0 \\
        0 & 0
    \end{pmatrix}
    \quad \text{and} \quad
    M_1 =
    \begin{pmatrix}
        0 & 0 \\
        0 & 1
    \end{pmatrix}.

> A POVM (Positive Operator-Valued Measure) is a set of positive operators that sum to the identity. 

Reviewer 2: ```suggestion

```

> Our function expects this set of operators to be a POVM because it checks if the sum of the operators 
> equals the identity, ensuring that the measurement outcomes are properly normalized.

Reviewer 2: ```suggestion
Our function expects this set of operators to be a POVM because it checks if the operators sum up to the identity, ensuring that the measurement outcomes are properly normalized.
```

    >>> from toqito.measurement_props import is_povm
    >>> import numpy as np
    >>> meas_1 = np.array([[1, 0], [0, 0]])
    >>> meas_2 = np.array([[0, 0], [0, 1]])
    >>> meas = [meas_1, meas_2]
    >>> is_povm(meas)
    True

Random POVM
^^^^^^^^^^^

> 
> A POVM (Positive Operator-Valued Measure) is a set of positive semidefinite operators {M₁, M₂, ..., Mₙ} 
> that sum to the identity matrix (∑ᵢMᵢ = I). Each operator Mᵢ corresponds to a possible measurement outcome, 
> and the probability of obtaining outcome i when measuring a quantum state ρ is given by Tr(Mᵢρ). 
> POVMs provide the most general description of quantum measurements allowed by quantum mechanics.

Reviewer 2: ```suggestion

```

Reviewer 2: I don't think you need to define a POVM here again since you have already specified it in the previous section.

> We may also use the :code:`random_povm` function from :code:`toqito`, and can verify that a

Reviewer 2: ```suggestion
We may also use :func:`.random_povm` function to randomly generate a POVM, and can verify that a
```

randomly generated set satisfies the criteria for being a POVM set.

    >>> from toqito.measurement_props import is_povm
    >>> from toqito.rand import random_povm
    >>> import numpy as np
    >>> dim, num_inputs, num_outputs = 2, 2, 2
    >>> measurements = random_povm(dim, num_inputs, num_outputs)
    >>> is_povm([measurements[:, :, 0, 0], measurements[:, :, 0, 1]])
    True

> Alternatively, the following matrices

Reviewer 2: ```suggestion
Alternatively, the following matrices do not constitute a POVM set.
```

.. math::
    M_0 =
    \begin{pmatrix}
        1 & 2 \\
        3 & 4
    \end{pmatrix}
    \quad \text{and} \quad
    M_1 =
    \begin{pmatrix}
        5 & 6 \\
        7 & 8
    \end{pmatrix},

> do not constitute a POVM set.

Reviewer 2: ```suggestion

```

    >>> from toqito.measurement_props import is_povm
    >>> import numpy as np
    >>> non_meas_1 = np.array([[1, 2], [3, 4]])
    >>> non_meas_2 = np.array([[5, 6], [7, 8]])
    >>> non_meas = [non_meas_1, non_meas_2]
    >>> is_povm(non_meas)
    False

> Measure

Reviewer 2: ```suggestion
Measurement Operators
```

^^^^^^^

Consider the following state:

.. math::
    u = \frac{1}{\sqrt{3}} e_0 + \sqrt{\frac{2}{3}} e_1

> where we define :math:`u u^* = \rho \in \text{D}(\mathcal{X})`.

Reviewer 2: ```suggestion
where we define :math:`u u^* = \rho \in \text{D}(\mathcal{X})` and :math:`e_0` and :math:`e_1` are the standard basis vectors.
```

> Let :math:`e_0` and :math:`e_1` be the standard basis vectors:

Reviewer 2: ```suggestion

```

.. math::
    e_0 = \begin{pmatrix} 1 \\ 0 \end{pmatrix} \quad \text{and} \quad e_1 = \begin{pmatrix} 0 \\ 1 \end{pmatrix}

> Define measurement operators

Reviewer 2: ```suggestion
The measurement operators are defined as shown below:
```

.. math::
    G_i = P^{-1/2} \left(p_i \rho_i\right) P^{-1/2} \quad \text{where} \quad P = \sum_{i=1}^n p_i \rho_i.

This measurement was initially defined in :cite:`Hughston_1993_Complete` and has found applications in quantum state discrimination tasks. 
While not always optimal, the PGM provides a reasonable measurement strategy that can be computed efficiently.

For example, consider the following trine states:

.. math::
    u_0 = |0\rangle, \quad
    u_1 = -\frac{1}{2}\left(|0\rangle + \sqrt{3}|1\rangle\right), \quad \text{and} \quad
    u_2 = -\frac{1}{2}\left(|0\rangle - \sqrt{3}|1\rangle\right).

.. code-block:: python

    >>> from toqito.states import trine
    >>> from toqito.measurements import pretty_good_measurement
    >>>
    >>> states = trine()
    >>> probs = [1 / 3, 1 / 3, 1 / 3]
    >>> pgm = pretty_good_measurement(states, probs)
    >>> pgm
    [array([[0.66666667, 0.        ],
           [0.        , 0.        ]]), array([[0.16666667, 0.28867513],
           [0.28867513, 0.5       ]]), array([[ 0.16666667, -0.28867513],
           [-0.28867513,  0.5       ]])]

Pretty Bad Measurement
^^^^^^^^^^^^^^^^^^^^^^

> Similarly, we can consider so-called "pretty bad measurement" (PBM) on the set of trine states :cite:`McIrvin_2024_Pretty`.

Reviewer 2: Since you define what a pretty good measurement is in the previous section, you should do the same for pbm..
